SimTracker/TrackAssociation/plugins/TrackTimeValueMapProducer.cc

```diff
@@ -129,120 +133,133 @@ void TrackTimeValueMapProducer::produce(edm::StreamID sid, edm::Event& evt, cons
     evt.getByToken(token,back);
   }
 
-  std::vector<float> generalTrackTimes, gsfTrackTimes;
+  std::vector<float> generalTrackTimes;
  
   //get track collections
   edm::Handle<edm::View<reco::Track> > TrackCollectionH;
   evt.getByToken(tracks_, TrackCollectionH);
   const edm::View<reco::Track>& TrackCollection = *TrackCollectionH;
 
-  edm::Handle<edm::View<reco::Track> > GsfTrackCollectionH;
-  evt.getByToken(gsfTracks_, GsfTrackCollectionH);
-  const edm::View<reco::Track>& GsfTrackCollection = *GsfTrackCollectionH;
-
   //get tracking particle collections
   edm::Handle<TrackingParticleCollection>  TPCollectionH;
   evt.getByToken(trackingParticles_, TPCollectionH);
-  //const TrackingParticleCollection&  TPCollection = *TPCollectionH;
+  
+  edm::Handle<std::vector<PileupSummaryInfo> > pileupSummaryH;
+  evt.getByToken(pileupSummaryInfo_, pileupSummaryH);
+  
+  //transient track builder
+  edm::ESHandle<TransientTrackBuilder> theB;
+  es.get<TransientTrackRecord>().get("TransientTrackBuilder",theB);
   
   // associate the reco tracks / gsf Tracks
-  std::vector<reco::RecoToSimCollection> associatedTracks, associatedTracksGsf;  
+  std::vector<reco::RecoToSimCollection> associatedTracks;  
   for( auto associator : associators ) {
     associatedTracks.emplace_back(associator->associateRecoToSim(TrackCollectionH, TPCollectionH));
-    associatedTracksGsf.emplace_back(associator->associateRecoToSim(GsfTrackCollectionH, TPCollectionH));
   }
   
 
-  calculateTrackTimes(TrackCollection, associatedTracks, generalTrackTimes);
-  calculateTrackTimes(GsfTrackCollection, associatedTracksGsf, gsfTrackTimes);
+  double sumSimTime = 0.;
+  double sumSimTimeSq = 0.;
+  int nsim = 0;
+  for (const PileupSummaryInfo &puinfo : *pileupSummaryH) {
+    if (puinfo.getBunchCrossing() == 0) {
+      for (const float &time : puinfo.getPU_times()) {
+        double simtime = time;
+        sumSimTime += simtime;
+        sumSimTimeSq += simtime*simtime;
+        ++nsim;
+      }
+      break;
+    }
+  }
+  
+  double meanSimTime = sumSimTime/double(nsim);
+  double varSimTime = sumSimTimeSq/double(nsim) - meanSimTime*meanSimTime;
+  double rmsSimTime = std::sqrt(std::max(0.1*0.1,varSimTime));
+
+  for( unsigned itk = 0; itk < TrackCollection.size(); ++itk ) {
+    const auto tkref = TrackCollection.refAt(itk);
+    reco::RecoToSimCollection::const_iterator track_tps = associatedTracks.back().end();
 
+    for( const auto& association : associatedTracks ) {
+      track_tps = association.find(tkref);
+      if( track_tps != association.end() ) break;
+    }
+
+    if (track_tps != associatedTracks.back().end() && track_tps->val.size() == 1) {
+      reco::TransientTrack tt = theB->build(*tkref);
+      float time = extractTrackVertexTime(*track_tps->val[0].first,tt);
+      generalTrackTimes.push_back(time);
+    }
+    else {
+      float rndtime = CLHEP::RandGauss::shoot(rng_engine, meanSimTime, rmsSimTime);
```

Comment:
[the functionality has been here before; I perhaps notice it more obviously only now]
this leads to non-reproducibility of RECO in multithreaded jobs and is technically in violation of the policy.
The seed should be reset to something based on event or track quantities.
It may be OK for now as long as the 4D vertex and track times are used only for checks or tests and do not contribute to the downstream reco.

I made #20621 to keep track of this issue.

Reply:
I get the appropriate random number engine based on the stream ID from the RNG service in the produce() method and pass it down. It should be OK... no?

Reply:
I guess not. It's not about streams. Events can be executed in different order, even if there is a continuous tracking of the seed/random engine state per stream.

Reply:
The RandomNumberGeneratorService associates random number sequences to streams. However, when running a job, the association of events to streams is not-reproducible (the same event may be assigned to different streams in different jobs). 

The standard for reconstruction, even before multi-threading, was to have the module seed the random number generator based on a property of the event.

Reply:
Gotcha. I'll try to make a PR soon to address this. Trying to attack the 4D vertex speed issue from a slightly different direction first.

Reply:
@lgray for a recent example, see #20240